controllers/cluster/capabilities.go

```diff
@@ -0,0 +1,168 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+
+	v1 "github.com/openshift/api/config/v1"
+	configv1 "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+)
+
+var (
+	versionRegex = regexp.MustCompile(`(\d+)\.(\d+)`)
+)
+
+type Capabilities struct {
+	IsOnOpenshift, HasMachineAPI bool
+}
+
+type ServerVersion struct {
+	OcpVersion   string
+	Major, Minor int
+}
+
+func NewCapabilities(config *rest.Config) (Capabilities, error) {
+	var err error
+	var onOpenshift, machineAPI bool
+
+	if onOpenshift, err = isOnOpenshift(config); err != nil {
+		return Capabilities{}, errors.Wrap(err, "failed to check if cluster is on OpenShift")
+	}
+
+	if onOpenshift {
+		if machineAPI, err = hasMachineAPI(config); err != nil {
+			return Capabilities{}, errors.Wrap(err, "failed to check machine API capability")
+		}
+	}
+
+	return Capabilities{
+		IsOnOpenshift: onOpenshift,
+		HasMachineAPI: machineAPI,
+	}, nil
+}
+
+// isOnOpenshift returns true if the cluster has the openshift config group
+func isOnOpenshift(config *rest.Config) (bool, error) {
+	dc, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return false, err
+	}
+	apiGroups, err := dc.ServerGroups()
+	kind := schema.GroupVersionKind{Group: "config.openshift.io", Version: "v1", Kind: "ClusterVersion"}
+	for _, apiGroup := range apiGroups.Groups {
+		for _, supportedVersion := range apiGroup.Versions {
+			if supportedVersion.GroupVersion == kind.GroupVersion().String() {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+// hasMachineAPI returns true it the cluster has the MachineAPI
+func hasMachineAPI(config *rest.Config) (bool, error) {
+	client, err := configv1.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to create cluster version client")
+	}
+	ocpVersion, err := getOpenshiftVersion(client)
+	if err != nil {
+		return false, errors.Wrap(err, "could not get OCP version")
+	}
+
+	if ocpVersion.Major == 4 && ocpVersion.Minor >= 14 {
+		return hasMachineAPICapability(client)
+	} else {
+		return hasMachineAPIGroup(config)
+	}
+}
+
+// getOpenshiftVersion returns the OpenShift version of the cluster
+func getOpenshiftVersion(client *configv1.ConfigV1Client) (*ServerVersion, error) {
+	clusterVersion, err := client.ClusterVersions().Get(context.TODO(), "version", metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get OCP version")
+	}
+
+	var ocpVersion string
+	for _, update := range clusterVersion.Status.History {
+		if update.State == v1.CompletedUpdate {
+			// obtain the version from the last completed update
+			ocpVersion = update.Version
+			break
+		}
+	}
+	if ocpVersion == "" {
+		return nil, errors.New("could not find up to date OCP version")
+	}
+
+	matches := versionRegex.FindStringSubmatch(ocpVersion)
+	if matches == nil || len(matches) != 3 {
```

Comment:
IIUC  len(matches) != 3 does not address future OCP versions (for example 5.1)

Reply:
the `matches` list is expected to include the full string, the major and the minor, so this should be correct.

Reply:
Maybe I'm missing something, but couldn't minor (major as well, but I guess that's a long time ahead) be also double digit ?
(i.e how would this work than with 4.16 for example ?)

Reply:
I'm more keen in dropping OCP version check and just look for MachineAPI Group. Keeping both implementations requires some complication we don't really need.

I'll take a note about the package @slintes linked though, I did not remember we used it :+1: 

I've just been affected by the inability to [kill your darlings](https://www.masterclass.com/articles/what-does-it-mean-to-kill-your-darlings) here :)

> Maybe I'm missing something, but couldn't minor (major as well, but I guess that's a long time ahead) be also double digit ?

double digits are taken into account by the `+` sign in the regex group `(\d+)`

Reply:
> I did not remember we used it

It's an indirect dependency only. I thought we use a semver lib somewhere directly though, but didn't find it, just that reference in go.mod 🙂